ring.go

```diff
@@ -220,25 +220,43 @@ type ringShards struct {
 }
 
 func newRingShards(opt *RingOptions) *ringShards {
-	shards := make(map[string]*ringShard, len(opt.Addrs))
+	shards, list := createRingShards(opt, opt.Addrs)
+
+	c := &ringShards{
+		opt: opt,
+
+		shards: shards,
+		list:   list,
+	}
+	c.rebalance()
+
+	return c
+}
+
+func createRingShards(opt *RingOptions, addrs map[string]string) (map[string]*ringShard, []*ringShard) {
+	shards := make(map[string]*ringShard, len(addrs))
 	list := make([]*ringShard, 0, len(shards))
 
-	for name, addr := range opt.Addrs {
+	for name, addr := range addrs {
 		shard := newRingShard(opt, name, addr)
 		shards[name] = shard
 
 		list = append(list, shard)
 	}
+	return shards, list
+}
 
-	c := &ringShards{
-		opt: opt,
+func (c *ringShards) SetAddrs(addrs map[string]string) {
+	shards, list := createRingShards(c.opt, addrs)
 
-		shards: shards,
-		list:   list,
+	c.mu.Lock()
+	if !c.closed {
+		c.shards = shards
+		c.list = list
```

Comment:
I think it should close old shards to prevent connection leak.
Also the common usecase seems to be addition/removal of a shard, e.g. `[a, b, c] -> [a, b, c, d]` - the proposed implementation would throw away perfectly fine shards a, b and c just to recreate them, it would also briefly double number of connections to each of them.
Besides some of existing shards could be down and re-creation would mark them up until next healthcheck.
Maybe the update should attempt to reuse existing shards if their name+addr are in the updated list.

Reply:
I agree with this. @szuecs what do you think?

Reply:
Yes of course!

Reply:
I made the change